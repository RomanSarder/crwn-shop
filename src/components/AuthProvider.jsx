import React, { useEffect, useState } from 'react'
import { auth, firestore, signOut, makeSignUpWithEmailAndPasswordFunction, makeSignInWithGoogleFunction } from '../firebase/utils'

export var AuthContext = React.createContext({
    currentUser: null
})

export default function AuthProvider({ children }) {
    var [currentUser, setCurrentUser] = useState(null)
    var [currentUserUid, setCurrentUserUid] = useState(null)
    
    useEffect(function subscribeToCurrentAuthenticatedUser () {
        var unsubscribe = auth.onAuthStateChanged(async function getUserProfile (user) {
            if (user) {
                try {
                    var userProfileSnapshot = await firestore
                        .collection('users')
                        .doc(user.uid)
                        .get()
    
                    var userProfileData = userProfileSnapshot.data()
                    if (currentUserUid !== user.uid) {
                        setCurrentUser(userProfileData)
                        setCurrentUserUid(user.uid)
                    }
                } catch (error) {
                    console.log(`Error while trying to get user profile:${error.message}`)
                }
            } else {
                setCurrentUserUid(null)
                setCurrentUser(null)
            }
        })

        return unsubscribe
    })

    var contextValue = {
        currentUser, 
        signOut, 
        makeSignUpFunction: makeSignUpWithEmailAndPasswordFunction,
        signInWithGoogle: makeSignInWithGoogleFunction()
    }

    return (
        <AuthContext.Provider value={contextValue}>
            {children}
        </AuthContext.Provider>
    )
}
