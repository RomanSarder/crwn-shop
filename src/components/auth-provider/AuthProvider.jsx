import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { 
    getAuthInstance, 
    getFirestoreInstance, 
    signOut, 
    makeSignUpWithEmailAndPasswordFunction, 
    signInWithGoogle, 
    signInWithEmailAndPassword 
} from '../../firebase/utils'
import { setUser } from '../../store/user/actions'
import { selectUserId } from '../../store/user/selectors'

export var AuthContext = React.createContext({
    signOut, 
    makeSignUpFunction: makeSignUpWithEmailAndPasswordFunction,
    signInWithGoogle,
    signInWithEmailAndPassword,
})

export function handleAuthStateChanged (dispatch, currentUserId) {
    return async function getUserProfile (user) {
        if (user) {
            if (currentUserId !== user.uid) {
                var firestore = await getFirestoreInstance()
                try {
                    var userProfileSnapshot = await firestore
                        .collection('users')
                        .doc(user.uid)
                        .get()
                    var userProfileData = userProfileSnapshot.data()
                    dispatch(setUser({ ...userProfileData, uid: userProfileSnapshot.id }))
                } catch (error) {
                    console.log(`Error while trying to get user profile:${error.message}`)
                }
            }
        } else {
            dispatch(setUser(null))
        }
    }
}

export default function AuthProvider({ children }) {

    var dispatch = useDispatch()
    var currentUserId = useSelector(selectUserId)
    
    
    useEffect(function subscribeToCurrentAuthenticatedUser () {
        var auth = getAuthInstance()
        var unsubscribe = auth.onAuthStateChanged(handleAuthStateChanged(dispatch, currentUserId))
        return unsubscribe
    }, [currentUserId])

    return (
        <AuthContext.Provider value={{
            signOut, 
            makeSignUpFunction: makeSignUpWithEmailAndPasswordFunction,
            signInWithGoogle,
            signInWithEmailAndPassword,
        }}>
            {children}
        </AuthContext.Provider>
    )
}
