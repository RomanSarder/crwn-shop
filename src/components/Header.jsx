import React, { useContext } from 'react'
import { shallowEqual, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import styled from 'styled-components'

import { ReactComponent as Logo } from '../assets/images/crown.svg'
import { getUser } from '../store/user/selectors'
import { AuthContext } from './AuthProvider'
import CartIcon from './CartIcon'

const StyledHeader = styled.div`
    display: flex;
    justify-content: space-between;

    .menu {
        display: flex;
        gap: 2rem;
    }

    a, span {
        text-decoration: none;
        color: black;
        font-size: 2.2rem;
        text-transform: uppercase;
        padding: 0;
    }

    button {
        background: none;
        font-weight: normal;
        display: inline-flex;
        align-items: flex-start;

        span {
            margin-top: -0.1rem;
        }
    }
`

export default function Header() {
    var { signOut } = useContext(AuthContext)
    var currentUser = useSelector(getUser, shallowEqual)

    return (
        <StyledHeader>
            <div className="logo-container">
                <Link to="/">
                    <Logo />
                </Link>
            </div>
            <div className="menu">
                <Link to="/shop">Shop</Link>
                <Link to="/contact">Contact</Link>
                { currentUser ?
                  <React.Fragment>
                      <span>{currentUser.displayName}</span>
                      <button onClick={() => signOut()}>
                          <span>Sign Out</span>
                      </button>
                      <CartIcon />
                  </React.Fragment> :
                  <Link to="/auth">Sign In</Link>   
                }
            </div>
        </StyledHeader>
    )
}
