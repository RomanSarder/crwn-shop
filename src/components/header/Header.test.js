import { render } from '@testing-library/react'
import 'jest-styled-components'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import configureStore from 'redux-mock-store'
import Header from './Header'
import { Provider } from 'react-redux'
import { AuthContext } from '../auth-provider/AuthProvider'
import userEvent from '@testing-library/user-event'

jest.mock('../cart-content/CartContent', () => () => (<div>Cart Content Mock</div>))
jest.mock('../cart-icon/CartIcon', () => () => (<div>Cart Icon Mock</div>))


var history = createMemoryHistory()
var signOutMock = jest.fn()

function renderHeader (store) {
    var renderResult = render(
        <Router history={history}>
            <Provider store={store}>
                <AuthContext.Provider value={{ signOut: signOutMock }}>
                    <Header />
                </AuthContext.Provider>
            </Provider>
        </Router>
    )

    return renderResult
}

it('<Header /> should display unauthenticated state', () => {
    var store = configureStore()({
        user: {
            currentUser: null
        },
        cart: {
            showCart: false
        }
    })

    var { container, getAllByRole, getByTestId } = renderHeader(store)

    expect(container.firstChild).toMatchSnapshot()
    var links = getAllByRole('link')
    expect(links.length).toEqual(4)
    expect(links[0]).toContainElement(getByTestId('logo-svg'))
    expect(links[0]).toHaveAttribute('href', '/')

    expect(links[1]).toHaveTextContent('Shop')
    expect(links[1]).toHaveAttribute('href', '/shop')
    expect(links[2]).toHaveTextContent('Contact')
    expect(links[2]).toHaveAttribute('href', '/contact')
    expect(links[3]).toHaveTextContent('Sign In')
    expect(links[3]).toHaveAttribute('href', '/auth')
})

it('<Header /> should display authenticated state', () => {
    const testDisplayName = 'Test User'
    var store = configureStore()({
        user: {
            currentUser: { displayName: testDisplayName }
        },
        cart: {
            showCart: true
        }
    })

    var { container, getAllByRole, getByRole, getByText } = renderHeader(store)
    expect(container.firstChild).toMatchSnapshot()
    var links = getAllByRole('link')
    expect(links.length).toEqual(3)
    expect(getByText('Test User')).toBeVisible()
    expect(getByRole('button')).toHaveTextContent('Sign Out')
    expect(getByText('Cart Icon Mock')).toBeVisible()
})

it('<Header /> have sign out button working', () => {
    const testDisplayName = 'Test User'
    var store = configureStore()({
        user: {
            currentUser: { displayName: testDisplayName }
        },
        cart: {
            showCart: false
        }
    })

    var { getByRole } = renderHeader(store)

    userEvent.click(getByRole('button'))

    expect(signOutMock).toHaveBeenCalledTimes(1)
})

it('<Header /> should show cart', () => {
    const testDisplayName = 'Test User'
    var store = configureStore()({
        user: {
            currentUser: { displayName: testDisplayName }
        },
        cart: {
            showCart: true
        }
    })

    var { getByText, container } = renderHeader(store)

    expect(container.firstChild).toMatchSnapshot()
    expect(getByText('Cart Content Mock')).toBeVisible()
})