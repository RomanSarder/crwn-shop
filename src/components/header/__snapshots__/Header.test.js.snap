// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should display authenticated state 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: relative;
  height: 3.8rem;
}

.c0 .logo {
  width: 4.2rem;
  height: 3.3rem;
}

.c0 .menu {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 a,
.c0 span {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: black;
  font-size: 1.6rem;
  text-transform: uppercase;
  padding: 0;
  text-align: center;
}

.c0 .sign-out-btn {
  background: none;
  font-weight: normal;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0 .sign-out-btn span {
  margin-top: -0.1rem;
}

@media (min-width:576px) {
  .c0 {
    height: 4.5rem;
  }

  .c0 a,
  .c0 span {
    font-size: 2.2rem;
    text-align: left;
  }

  .c0 .logo {
    height: 3.9rem;
    width: 5rem;
  }

  .c0 .menu {
    -webkit-align-items: flex-start;
    -webkit-box-align: flex-start;
    -ms-flex-align: flex-start;
    align-items: flex-start;
  }
}

<div
  class="c0"
>
  <div
    class="logo-container"
  >
    <a
      href="/"
    >
      <svg
        class="logo"
        data-testid="logo-svg"
      >
        crown.svg
      </svg>
    </a>
  </div>
  <div
    class="menu"
  >
    <a
      href="/shop"
    >
      Shop
    </a>
    <a
      href="/contact"
    >
      Contact
    </a>
    <span>
      Test User
    </span>
    <button
      class="sign-out-btn"
    >
      <span>
        Sign Out
      </span>
    </button>
    <div>
      Cart Icon Mock
    </div>
  </div>
  <div>
    Cart Content Mock
  </div>
</div>
`;

exports[`<Header /> should display unauthenticated state 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: relative;
  height: 3.8rem;
}

.c0 .logo {
  width: 4.2rem;
  height: 3.3rem;
}

.c0 .menu {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 a,
.c0 span {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: black;
  font-size: 1.6rem;
  text-transform: uppercase;
  padding: 0;
  text-align: center;
}

.c0 .sign-out-btn {
  background: none;
  font-weight: normal;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0 .sign-out-btn span {
  margin-top: -0.1rem;
}

@media (min-width:576px) {
  .c0 {
    height: 4.5rem;
  }

  .c0 a,
  .c0 span {
    font-size: 2.2rem;
    text-align: left;
  }

  .c0 .logo {
    height: 3.9rem;
    width: 5rem;
  }

  .c0 .menu {
    -webkit-align-items: flex-start;
    -webkit-box-align: flex-start;
    -ms-flex-align: flex-start;
    align-items: flex-start;
  }
}

<div
  class="c0"
>
  <div
    class="logo-container"
  >
    <a
      href="/"
    >
      <svg
        class="logo"
        data-testid="logo-svg"
      >
        crown.svg
      </svg>
    </a>
  </div>
  <div
    class="menu"
  >
    <a
      href="/shop"
    >
      Shop
    </a>
    <a
      href="/contact"
    >
      Contact
    </a>
    <a
      href="/auth"
    >
      Sign In
    </a>
  </div>
</div>
`;

exports[`<Header /> should show cart 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: relative;
  height: 3.8rem;
}

.c0 .logo {
  width: 4.2rem;
  height: 3.3rem;
}

.c0 .menu {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 a,
.c0 span {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: black;
  font-size: 1.6rem;
  text-transform: uppercase;
  padding: 0;
  text-align: center;
}

.c0 .sign-out-btn {
  background: none;
  font-weight: normal;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0 .sign-out-btn span {
  margin-top: -0.1rem;
}

@media (min-width:576px) {
  .c0 {
    height: 4.5rem;
  }

  .c0 a,
  .c0 span {
    font-size: 2.2rem;
    text-align: left;
  }

  .c0 .logo {
    height: 3.9rem;
    width: 5rem;
  }

  .c0 .menu {
    -webkit-align-items: flex-start;
    -webkit-box-align: flex-start;
    -ms-flex-align: flex-start;
    align-items: flex-start;
  }
}

<div
  class="c0"
>
  <div
    class="logo-container"
  >
    <a
      href="/"
    >
      <svg
        class="logo"
        data-testid="logo-svg"
      >
        crown.svg
      </svg>
    </a>
  </div>
  <div
    class="menu"
  >
    <a
      href="/shop"
    >
      Shop
    </a>
    <a
      href="/contact"
    >
      Contact
    </a>
    <span>
      Test User
    </span>
    <button
      class="sign-out-btn"
    >
      <span>
        Sign Out
      </span>
    </button>
    <div>
      Cart Icon Mock
    </div>
  </div>
  <div>
    Cart Content Mock
  </div>
</div>
`;
